google/cloud/opentelemetry/internal/monitored_resource.cc

```diff
@@ -0,0 +1,222 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/opentelemetry/internal/monitored_resource.h"
+#include "google/cloud/version.h"
+#include "absl/types/optional.h"
+#include "absl/types/variant.h"
+#include <opentelemetry/common/attribute_value.h>
+#include <opentelemetry/sdk/resource/resource.h>
+#include <opentelemetry/sdk/resource/semantic_conventions.h>
+#include <unordered_map>
+
+namespace google {
+namespace cloud {
+namespace otel_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace {
+
+namespace sc = opentelemetry::sdk::resource::SemanticConventions;
+
+std::string AsString(
+    opentelemetry::sdk::common::OwnedAttributeValue attribute) {
+  if (absl::holds_alternative<std::string>(attribute)) {
+    return absl::get<std::string>(attribute);
+  }
+  // TODO(#11775) - convert other attribute types to strings
+  // Note that our resource detector only uses string attributes.
+  return {};
+}
+
+struct OTelKeyMatch {
+  std::vector<std::string> otel_keys;
+  absl::optional<std::string> fallback = absl::nullopt;
+};
+
+class MonitoredResourceProvider {
+ public:
+  MonitoredResourceProvider(
+      std::string type, std::unordered_map<std::string, OTelKeyMatch> lookup)
+      : type_(std::move(type)), lookup_(std::move(lookup)) {}
+
+  MonitoredResource Process(
+      opentelemetry::sdk::resource::ResourceAttributes const& attributes) {
+    MonitoredResource mr;
+    mr.type = type_;
+    for (auto const& kv : lookup_) {
+      auto found = false;
+      for (auto const& otel_key : kv.second.otel_keys) {
+        auto p = attributes.find(otel_key);
+        if (p != attributes.end()) {
+          found = true;
+          mr.labels[kv.first] = AsString(p->second);
+          break;
+        }
+      }
+      if (!found && kv.second.fallback)
```

Comment:
nit: need braces since it is too long for a one liner.

Reply:
N/A